pkg/queue/cluster_queue_interface.go

```diff
@@ -0,0 +1,80 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package queue
+
+import (
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+	kueue "sigs.k8s.io/kueue/api/v1alpha1"
+	"sigs.k8s.io/kueue/pkg/workload"
+)
+
+// ClusterQueue is an interface for a cluster queue to store workloads waiting
+// to be scheduled.
+type ClusterQueue interface {
+	// Update updates the properties of this ClusterQueue.
+	Update(*kueue.ClusterQueue)
+
+	// AddFromQueue pushes all workloads belonging to this queue to
+	// the ClusterQueue. If one of the workloads is already present,
+	// return false. Otherwise return true.
+	AddFromQueue(*Queue) bool
+	// DeleteFromQueue removes all workloads belonging to this queue from
+	// the ClusterQueue.
+	DeleteFromQueue(*Queue)
+
+	// PushIfNotPresent pushes the workload to ClusterQueue.
+	// If the workload is already present, return false. Otherwise return true.
+	PushIfNotPresent(*workload.Info) bool
+	// PushOrUpdate pushes the workload to ClusterQueue.
+	// If the workload is already present, update with the new one.
+	PushOrUpdate(*kueue.QueuedWorkload)
+	// Delete removes the workload from ClusterQueue.
+	Delete(*kueue.QueuedWorkload)
+	// Pop removes the head of the queue and returns it. It returns nil if the
+	// queue is empty.
+	Pop() *workload.Info
+
+	// RequeueWorkload pushes the workload back to ClusterQueue after scheduling failure.
+	RequeueWorkload(*workload.Info) (bool, error)
+	// AddInadmissibleIfNotPresent inserts a workload that cannot be admitted into
+	// the inadmissibleWorkloads in ClusterQueue, unless it is already in the queue.
+	AddInadmissibleIfNotPresent(*workload.Info) error
+	// QueueAllInadmissibleWorkloads moves all workloads from inadmissibleWorkloads to heap.
+	QueueAllInadmissibleWorkloads()
+
+	// Pending returns the size of pending workloads.
+	Pending() int32
+	// Dump produces a dump of the current workloads in the heap of
+	// this ClusterQueue. It returns false if the queue is empty.
+	// Otherwise return true.
+	Dump() (sets.String, bool)
```

Comment:
For the future, I think we can use a similar interface to support the following UX experience: an admin or even user may want to see where the job is in the queue. So we may want to rename it to something that indicates more than "testing/debugging", but I am not suggesting we rename it now.

Reply:
> where the job is in the queue

is it the exact position in the queue, e.g. the 5th one in the queue？If so, we can return a `slice` instead of `sets` in the future?

Reply:
something like that, yeah we will need to change to an ordered slice.